presto-main/src/main/java/com/facebook/presto/server/remotetask/WorkerTaskStatusFetcher.java

```diff
@@ -0,0 +1,314 @@
+    private final Codec<Map<TaskId, TaskStatus>> taskListStatusCodec;
+    private final Duration refreshMaxWait;
+    private final Executor executor;
+    private final HttpClient httpClient;
+    private final BatchRequestErrorTracker errorTracker;
+    private final RemoteTaskStats stats;
+    private final boolean isBinaryTransportEnabled;
+    private final String sessionId;
+
+    private final AtomicLong currentRequestStartNanos = new AtomicLong();
+
+    @GuardedBy("this")
+    private boolean running;
+
+    @GuardedBy("this")
+    private ListenableFuture<BaseResponse<Map<TaskId, TaskStatus>>> future;
+
+    public WorkerTaskStatusFetcher(
+            String worker,
+            Consumer<Throwable> onFail,
+            Duration refreshMaxWait,
+            Codec<Map<TaskId, TaskStatus>> taskListStatusCodec,
+            Executor executor,
+            HttpClient httpClient,
+            Duration maxErrorDuration,
+            ScheduledExecutorService errorScheduledExecutor,
+            RemoteTaskStats stats,
+            boolean isBinaryTransportEnabled,
+            String sessionId)
+    {
+        this.worker = worker;
+        idTaskMap = new ConcurrentHashMap<>();
+
+        this.onFail = requireNonNull(onFail, "onFail is null");
+
+        this.refreshMaxWait = requireNonNull(refreshMaxWait, "refreshMaxWait is null");
+        this.taskListStatusCodec = requireNonNull(taskListStatusCodec, "taskStatusCodec is null");
+
+        this.executor = requireNonNull(executor, "executor is null");
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+
+        this.errorTracker = new BatchRequestErrorTracker(
+                Collections.list(idTaskMap.keys()),
+                getWorkerURI(),
+                maxErrorDuration,
+                errorScheduledExecutor,
+                "getting task status");
+        this.stats = requireNonNull(stats, "stats is null");
+        this.isBinaryTransportEnabled = isBinaryTransportEnabled;
+
+        this.sessionId = sessionId;
+    }
+
+    public void addTask(ContinuousBatchTaskStatusFetcher.Task task)
+    {
+        idTaskMap.put(
+                task.taskId,
+                new Pair<ContinuousBatchTaskStatusFetcher.Task, StateMachine<TaskStatus>>(
+                        task,
+                        new StateMachine(
+                                "task-" + task.taskId, executor, task.taskStatus)));
+    }
+
+    TaskStatus getTaskStatus(TaskId taskId)
+    {
+        return idTaskMap.get(taskId).getKey().taskStatus.get();
+    }
+
+    private URI getWorkerURI()
+    {
+        URI uri;
+        try {
+            uri = uriBuilderFrom(new URI(worker)).appendPath("v1/task/status?session_id=" + sessionId).build();
+        }
+        catch (URISyntaxException e) {
+            uri = null;
+        }
+        return uri;
+    }
+
+    public synchronized void scheduleNextRequest()
+    {
+        // stopped or done?
+        if (!running) {
```

Comment:
let's just remove this condition, not needed right now. With this condition it will not run ever, i.e. running = false and this condition is always true.